contracts/helpers/Rebalancer.sol

```diff
@@ -0,0 +1,152 @@
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Initializes the immutable variables of the contract, namely `_accessControlManager` and `_transmuter`
+    constructor(IAccessControlManager _accessControlManager, ITransmuter _transmuter) {
+        if (address(_accessControlManager) == address(0)) revert ZeroAddress();
+        accessControlManager = _accessControlManager;
+        transmuter = _transmuter;
+        agToken = address(_transmuter.agToken());
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+                                                 REBALANCING FUNCTIONS                                              
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Swaps `tokenIn` for `tokenOut` through an intermediary agToken mint from `tokenIn` and
+    /// burn to `tokenOut`. Eventually, this transaction may be sponsored with an additional amount
+    /// of `tokenOut` based on the `orders` set by governance
+    /// @param amountIn Amount of `tokenIn` to bring for the rebalancing
+    /// @param amountOutMin Minimum amount of `tokenOut`that must be obtained from the swap
+    /// @param to Address to which `tokenOut` must be sent
+    /// @param deadline Timestamp before which this transaction must be included
+    /// @dev Contrarily to what is done in the Transmuter contract, here neither of `tokenIn` or `tokenOut`
+    /// should be an `agToken`
+    function swapExactInput(
+        uint256 amountIn,
+        uint256 amountOutMin,
+        address tokenIn,
+        address tokenOut,
+        address to,
+        uint256 deadline
+    ) external returns (uint256 amountOut) {
+        IERC20(tokenIn).safeTransferFrom(msg.sender, address(this), amountIn);
+        // Dealing with the allowance of the rebalancer to the Transmuter: this allowance is made infinite by default
+        uint256 allowance = IERC20(tokenIn).allowance(address(this), address(transmuter));
+        if (allowance < amountIn)
+            IERC20(tokenIn).safeIncreaseAllowance(address(transmuter), type(uint256).max - allowance);
+        // First, mint agToken from `tokenIn`
+        uint256 amountAgToken = transmuter.swapExactInput(
+            amountIn,
+            0,
+            tokenIn,
+            agToken,
+            address(this),
+            block.timestamp
+        );
+        // Then, burn the minted agToken to `tokenOut`
+        amountOut = transmuter.swapExactInput(amountAgToken, 0, agToken, tokenOut, to, deadline);
+        // Based on the `amountOut` obtained, checking whether this is eligible to a subsidy
+        uint256 subsidy = _computeSubsidy(tokenIn, tokenOut, amountOut);
```

Comment:
I find it odd to have to pay the premium in tokenOut and not agToken
Also it will be a pain to pay in outToken as most of the time outToken will be a KYC token, so hard to get.

If we pay in agToken and based on `amountAgToken`, the risk shouldn't be very large and if we are in this setup - in which we want to buy outToken - fees should be null on the transmuter itself and so the only difference will be the oracle rate. If not at par, depeg or whatever, then we may lose our premiums (or badly allocate it) but it is a one time thing.

We can also send it in inToken, and it should also be ok for the same argument of the above paragraph

Reply:
Based on current dynamics (when we're growing), most of the time, outToken will be EUROC which we do control in reserves.

Paying in InToken makes less sense to me if you're considering that you're growing.

On the tokens being hard to get, in any case if we have the agToken we can always burn for some of the outToken, so already perform a small rebalancing without having to pay the market makers.

I think the agToken can still be an idea to explore, I just fear it makes it harder at the interface level, and from a user perspective for the market makers